File: Days 61 - 70/Day 67/LFUCache.cpp
#include <iostream>
#include <map>
#include <memory>
#include <stdexcept>
#include <string>
#include <unordered_map>

class LFUCache
{
private:
	struct Node
	{
	public:
		std::string key;
		std::string value;

		std::shared_ptr<Node> next;
		std::shared_ptr<Node> previous;

	public:
		Node(const std::string& key, const std::string& value)
			: key(key), value(value)
		{ }
	};

	class LinkedList
	{
	private:
		std::shared_ptr<Node> head = nullptr;
		std::shared_ptr<Node> tail = nullptr;

		unsigned int size = 0;

	public:
		void Add(std::shared_ptr<Node> node)
		{
			if (head == nullptr)
			{
				head = node;
			}
			else
			{
				tail->next = node;
				node->previous = tail;
			}

			tail = node;
			size++;
		}

		void Remove(std::shared_ptr<Node> node)
		{

			if (node->next == nullptr)
			{
				tail = node->previous;
			}
			else
			{
				node->next->previous = node->previous;
			}

			if (head->key == node->key)
			{
				head = node->next;
			}
			else
			{
				node->previous->next = node->next;
			}

			size--;
		}

		inline std::shared_ptr<Node> GetHead() { return head; }
		inline unsigned int GetSize() { return size; }
	};

	std::unordered_map<std::string, std::shared_ptr<Node>> values;
	std::unordered_map<std::string, unsigned int> counts;
	std::map<unsigned int, LinkedList> frequencies;

	const int limit;

public:
	explicit LFUCache(int limit)
		: limit(limit)
	{ }

	std::string Get(const std::string& key)
	{
		if (values.find(key) == values.end())
		{
			throw std::out_of_range("Key not in cache.");
		}

		std::shared_ptr<Node> node = values[key];

		int frequency = counts[key];
		frequencies[frequency].Remove(node);
		RemoveEmptyList(frequency);

		if (frequencies.find(frequency + 1) == frequencies.end())
		{
			frequencies[frequency + 1].Add(node);
		}

		counts[key] = frequency + 1;

		return values[key]->value;
	}

	void Set(std::string key, std::string value)
	{
		if (values.find(key) == values.end())
		{
			std::shared_ptr<Node> node = std::make_shared<Node>(key, value);

			if (values.size() == limit)
			{
				unsigned int minCount = (*frequencies.begin()).first;
				std::shared_ptr<Node> nodeTodelete = frequencies[minCount].GetHead();
				frequencies[minCount].Remove(nodeTodelete);

				std::string keyToDelete = nodeTodelete->key;
				RemoveEmptyList(minCount);

				values.erase(keyToDelete);
				counts.erase(keyToDelete);
			}

			values[key] = node;
			counts[key] = 1;

			if (frequencies.find(1) == frequencies.end())
			{
				frequencies[1].Add(node);
			}
		}
	}

private:
	void RemoveEmptyList(unsigned int frequency)
	{
		if (frequencies[frequency].GetSize() == 0)
		{
			frequencies.erase(frequency);
		}
	}
};

int main(int argc, char* argv[])
{
	LFUCache cache(5);
	cache.Set("1", "1");
	cache.Set("2", "2");
	cache.Set("3", "3");
	cache.Set("4", "4");
	cache.Set("5", "5");

	cache.Set("2", "2");
	cache.Set("3", "3");
	cache.Set("4", "4");
	cache.Set("5", "5");

	cache.Set("3", "3");
	cache.Set("4", "4");
	cache.Set("5", "5");

	cache.Set("4", "4");
	cache.Set("5", "5");

	cache.Set("5", "5");

	cache.Set("6", "6");

	try
	{
		std::cout << cache.Get("1") << "\n";
	}
	catch (const std::out_of_range&)
	{
		std::cout << "1 has been removed from the cache.\n";
	}

	std::cin.get();

	return 0;
}

File: Days 061 - 070/Day 67/LFUCache.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace DailyCodingProblem
{
	internal class LFUCache
	{
		private class Node
		{
			public string Key { get; set; }
			public string Value { get; set; }

			public Node Next { get; set; }
			public Node Previous { get; set; }

			public Node(string key, string value)
			{
				this.Key = key;
				this.Value = value;
			}
		}

		private class LinkedList
		{
			public  Node Head { get; private set; } = null;
			private Node tail = null;

			public int Size { get; private set; } = 0;

			public void	Add(Node node)
			{
				if (Head == null)
				{
					Head = node;
				}
				else
				{
					tail.Next = node;
					node.Previous = tail;
				}

				tail = node;
				Size++;
			}

			public void Remove(Node node)
			{
				if (node.Next == null)
				{
					tail = node.Previous;
				}
				else
				{
					node.Next.Previous = node.Previous;
				}

				if (Head.Key == node.Key)
				{
					Head = node.Next;
				}
				else
				{
					node.Previous.Next = node.Next;
				}

				Size++;
			}
		}

		private Dictionary<string, Node> values = new Dictionary<string, Node>();
		private Dictionary<string, int> counts = new Dictionary<string, int>();
		private SortedDictionary<int, LinkedList> frequencies = new SortedDictionary<int, LinkedList>();

		private readonly int limit;

		public LFUCache(int limit)
		{
			this.limit = limit;
		}

		public string Get(string key)
		{
			if (!values.ContainsKey(key))
			{
				throw new ArgumentOutOfRangeException("Key not in cache.");
			}

			Node node = values[key];
			int frequency = counts[key];
			frequencies[frequency].Remove(node);
			RemoveEmptyList(frequency);

			if (!frequencies.ContainsKey(frequency + 1))
			{
				frequencies.Add(frequency + 1, new LinkedList());
				frequencies[frequency + 1].Add(node);
			}

			counts[key] = frequency + 1;

			return values[key].Value;
		}

		public void Set(string key, string value)
		{
			if (!values.ContainsKey(key))
			{
				Node node = new Node(key, value);

				if (value.Length == limit)
				{
					int minCount = frequencies.First().Key;
					Node nodeToDelete = frequencies[minCount].Head;
					frequencies[minCount].Remove(nodeToDelete);

					string keyToDelete = nodeToDelete.Key;
					RemoveEmptyList(minCount);

					values.Remove(keyToDelete);
					counts.Remove(keyToDelete);
				}

				values[key] = node;
				counts[key] = 1;

				if (!frequencies.ContainsKey(1))
				{
					frequencies.Add(1, new LinkedList());
					frequencies[1].Add(node);
				}
			}
		}

		private void RemoveEmptyList(int frequency)
		{
			if (frequencies[frequency].Size == 0)
			{
				frequencies.Remove(frequency);
			}
		}
	}

	internal class Day67
	{
		private static int Main(string[] args)
		{
			LFUCache cache = new LFUCache(5);
			cache.Set("1", "1");
			cache.Set("2", "2");
			cache.Set("3", "3");
			cache.Set("4", "4");
			cache.Set("5", "5");

			cache.Set("2", "2");
			cache.Set("3", "3");
			cache.Set("4", "4");
			cache.Set("5", "5");

			cache.Set("3", "3");
			cache.Set("4", "4");
			cache.Set("5", "5");

			cache.Set("4", "4");
			cache.Set("5", "5");

			cache.Set("5", "5");

			cache.Set("6", "6");

			try
			{
				Console.WriteLine(cache.Get("1"));
			}
			catch (ArgumentOutOfRangeException)
	{
				Console.WriteLine("1 has been removed from the cache.");
			}

			Console.ReadLine();

			return 0;
		}
	}
}